core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/update/userdata/UserDataUpdateActions.java

```diff
@@ -0,0 +1,127 @@
+import com.sequenceiq.cloudbreak.core.flow2.stack.StackContext;
+import com.sequenceiq.cloudbreak.domain.Resource;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.reactor.api.event.StackEvent;
+import com.sequenceiq.cloudbreak.reactor.api.event.stack.userdata.UserDataUpdateFailed;
+import com.sequenceiq.cloudbreak.reactor.api.event.stack.userdata.UserDataUpdateOnProviderRequest;
+import com.sequenceiq.cloudbreak.reactor.api.event.stack.userdata.UserDataUpdateOnProviderResult;
+import com.sequenceiq.cloudbreak.reactor.api.event.stack.userdata.UserDataUpdateRequest;
+import com.sequenceiq.cloudbreak.reactor.api.event.stack.userdata.UserDataUpdateSuccess;
+import com.sequenceiq.cloudbreak.service.image.ImageService;
+import com.sequenceiq.cloudbreak.service.resource.ResourceService;
+import com.sequenceiq.common.api.type.CommonStatus;
+import com.sequenceiq.common.api.type.InstanceGroupType;
+import com.sequenceiq.common.api.type.ResourceType;
+
+@Configuration
+public class UserDataUpdateActions {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserDataUpdateActions.class);
+
+    @Inject
+    private ResourceService resourceService;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private ResourceToCloudResourceConverter resourceToCloudResourceConverter;
+
+    @Bean(name = "UPDATE_USERDATA_STATE")
+    public AbstractUserDataUpdateAction<?> updateUserData() {
+        return new AbstractUserDataUpdateAction<>(UserDataUpdateRequest.class) {
+            @Override
+            protected void doExecute(StackContext context, UserDataUpdateRequest payload, Map<Object, Object> variables) throws Exception {
+                LOGGER.info("Recreate userdata for new freeipa instances");
+                sendEvent(context, new UserDataUpdateRequest(context.getStack().getId(), payload.getOldTunnel()));
+            }
+
+            @Override
+            protected Object getFailurePayload(UserDataUpdateRequest payload, Optional<StackContext> flowContext, Exception ex) {
+                return new UserDataUpdateFailed(payload.getResourceId(), ex);
+            }
+        };
+    }
+
+    @Bean(name = "UPDATE_USERDATA_ON_PROVIDER_STATE")
+    public AbstractUserDataUpdateAction<?> updateUserDataOnProviderSide() {
+        return new AbstractUserDataUpdateAction<>(UserDataUpdateSuccess.class) {
+            @Override
+            protected void doExecute(StackContext context, UserDataUpdateSuccess payload, Map<Object, Object> variables) {
+                sendEvent(context);
+            }
+
+            @Override
+            protected Selectable createRequest(StackContext context) {
+                Stack stack = context.getStack();
+                String userData = null;
+                try {
+                    userData = imageService.getImage(stack.getId()).getUserdata().get(InstanceGroupType.GATEWAY);
```

Comment:
Why do we store the user data in the payload?
It would be adequate to extract it in the handler (and update, of course).

Reply:
I copied it from the freeipa solution